File: iot_driver_copilot/temperature_and_humidity_sensor/ModbusRTU.java
import java.io.*;
import java.time.Instant;
import java.util.Arrays;

class ModbusRTU implements Closeable {
    private final String devicePath;
    private FileInputStream in;
    private FileOutputStream out;

    ModbusRTU(String devicePath) {
        this.devicePath = devicePath;
    }

    public synchronized void open() throws IOException {
        if (isOpen()) return;
        try {
            this.in = new FileInputStream(devicePath);
            this.out = new FileOutputStream(devicePath);
        } catch (FileNotFoundException e) {
            throw new IOException("Cannot open serial port: " + devicePath + ": " + e.getMessage(), e);
        }
    }

    public synchronized boolean isOpen() {
        return in != null && out != null;
    }

    public synchronized void close() {
        try { if (in != null) in.close(); } catch (Exception ignored) {}
        try { if (out != null) out.close(); } catch (Exception ignored) {}
        in = null; out = null;
    }

    public synchronized int[] readRegisters(int slaveId, int function, int startAddr, int count, int timeoutMs) throws IOException {
        if (!isOpen()) throw new IOException("serial not open");
        if (count <= 0 || count > 125) throw new IOException("invalid register count");

        // Build request ADU
        byte[] req = new byte[8];
        req[0] = (byte) (slaveId & 0xFF);
        req[1] = (byte) (function & 0xFF);
        req[2] = (byte) ((startAddr >> 8) & 0xFF);
        req[3] = (byte) (startAddr & 0xFF);
        req[4] = (byte) ((count >> 8) & 0xFF);
        req[5] = (byte) (count & 0xFF);
        int crc = crc16(req, 0, 6);
        req[6] = (byte) (crc & 0xFF);       // CRC Lo
        req[7] = (byte) ((crc >> 8) & 0xFF); // CRC Hi

        // Purge any stale bytes
        purgeInput();

        // send
        out.write(req);
        out.flush();

        // receive
        int expectedDataBytes = count * 2;
        int minLen = 3 + expectedDataBytes + 2; // addr + func + byteCount + data + crc
        byte[] buf = new byte[minLen];
        int read = 0;
        long deadline = System.currentTimeMillis() + timeoutMs;
        while (System.currentTimeMillis() < deadline && read < minLen) {
            int avail = 0;
            try { avail = in.available(); } catch (IOException ignored) {}
            if (avail <= 0) {
                try { Thread.sleep(5); } catch (InterruptedException ie) { Thread.currentThread().interrupt(); }
                continue;
            }
            int r = in.read(buf, read, Math.min(avail, minLen - read));
            if (r > 0) read += r;
        }

        if (read < minLen) {
            throw new IOException("timeout waiting for response ("+read+"/"+minLen+") at "+ Instant.now());
        }

        // Validate
        int p = 0;
        int respAddr = buf[p++] & 0xFF;
        int respFunc = buf[p++] & 0xFF;
        int byteCount = buf[p++] & 0xFF;
        if (respAddr != (slaveId & 0xFF)) throw new IOException("slave id mismatch resp="+respAddr);
        if (respFunc != (function & 0xFF)) throw new IOException("function mismatch resp="+respFunc);
        if (byteCount != expectedDataBytes) throw new IOException("byteCount mismatch resp="+byteCount+" expected="+expectedDataBytes);
        int dataEnd = 3 + byteCount;
        int crcResp = ((buf[dataEnd+1] & 0xFF) << 8) | (buf[dataEnd] & 0xFF);
        int crcCalc = crc16(buf, 0, dataEnd);
        if (crcResp != crcCalc) throw new IOException("CRC mismatch");

        int[] regs = new int[count];
        for (int i = 0; i < count; i++) {
            int hi = buf[3 + i*2] & 0xFF;
            int lo = buf[3 + i*2 + 1] & 0xFF;
            regs[i] = (hi << 8) | lo;
        }
        return regs;
    }

    private void purgeInput() {
        try {
            int avail = in.available();
            if (avail > 0) {
                byte[] tmp = new byte[Math.min(avail, 4096)];
                while (in.available() > 0) {
                    int r = in.read(tmp, 0, Math.min(tmp.length, in.available()));
                    if (r <= 0) break;
                }
            }
        } catch (IOException ignored) {}
    }

    static int crc16(byte[] data, int off, int len) {
        int crc = 0xFFFF;
        for (int i = off; i < off + len; i++) {
            crc ^= (data[i] & 0xFF);
            for (int j = 0; j < 8; j++) {
                if ((crc & 0x0001) != 0) {
                    crc >>= 1;
                    crc ^= 0xA001;
                } else {
                    crc >>= 1;
                }
            }
        }
        return crc & 0xFFFF;
    }
}

File: iot_driver_copilot/industrial_test_2/modbus_rtu.cpp
#include "modbus_rtu.h"
#include <cstring>

ModbusRTU::ModbusRTU(SerialPort* port, int timeout_ms) : port_(port), timeout_ms_(timeout_ms), last_error_("") {}

uint16_t ModbusRTU::crc16(const uint8_t* data, size_t len) {
    uint16_t crc = 0xFFFF;
    for (size_t pos = 0; pos < len; pos++) {
        crc ^= (uint16_t)data[pos];
        for (int i = 0; i < 8; i++) {
            if (crc & 1)
                crc = (crc >> 1) ^ 0xA001;
            else
                crc >>= 1;
        }
    }
    return crc;
}

bool ModbusRTU::readHoldingRegisters(uint8_t slave_id, uint16_t address, uint16_t quantity, std::vector<uint16_t>& out) {
    out.clear();
    if (!port_ || !port_->isOpen()) { last_error_ = "serial not open"; return false; }
    if (quantity < 1 || quantity > 125) { last_error_ = "invalid quantity"; return false; }

    uint8_t req[8];
    req[0] = slave_id;
    req[1] = 0x03; // function code: read holding registers
    req[2] = (uint8_t)((address >> 8) & 0xFF);
    req[3] = (uint8_t)(address & 0xFF);
    req[4] = (uint8_t)((quantity >> 8) & 0xFF);
    req[5] = (uint8_t)(quantity & 0xFF);
    uint16_t crc = crc16(req, 6);
    req[6] = (uint8_t)(crc & 0xFF);       // CRC low
    req[7] = (uint8_t)((crc >> 8) & 0xFF); // CRC high

    // Flush any pending I/O before sending a new request
    // Not all platforms need an explicit flush here; using tcflush in open.

    if (!port_->writeAll(req, sizeof(req))) {
        last_error_ = "write failed";
        return false;
    }

    // Expected response: addr(1) func(1) bytecount(1) data(2*quantity) crc(2)
    size_t expected = 5 + 2 * quantity;
    std::vector<uint8_t> resp(expected);
    if (!port_->readExact(resp.data(), resp.size(), timeout_ms_)) {
        last_error_ = "read timeout";
        return false;
    }

    // Validate address and function
    if (resp[0] != slave_id) { last_error_ = "mismatched slave id"; return false; }
    if (resp[1] == (0x80 | 0x03)) { last_error_ = "modbus exception"; return false; }
    if (resp[1] != 0x03) { last_error_ = "unexpected function"; return false; }

    uint8_t byte_count = resp[2];
    if (byte_count != 2 * quantity) { last_error_ = "unexpected byte count"; return false; }

    // CRC check
    uint16_t rcrc = (uint16_t)resp[expected - 1] << 8 | resp[expected - 2];
    uint16_t ccrc = crc16(resp.data(), expected - 2);
    if (rcrc != ccrc) {
        last_error_ = "crc mismatch";
        return false;
    }

    out.reserve(quantity);
    for (size_t i = 0; i < quantity; ++i) {
        uint16_t hi = resp[3 + i * 2];
        uint16_t lo = resp[3 + i * 2 + 1];
        out.push_back((uint16_t)((hi << 8) | lo));
    }

    last_error_.clear();
    return true;
}
